confluent_kafka/schema_registry/schema.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import json
+
+PRIMITIVES = {'string', 'bytes', 'boolean', 'null',
+              'int', 'double', 'float', 'long'}
+
+
+class Schema(object):
+    """
+    An unregistered Schema.
+
+    Args:
+        schema (str): String representation of the schema.
+        references [Schema]: Schema's referenced by this one.
+        schema_type (str): The schema type: AVRO, PROTOBUF or JSON.
+
+    """
+    __slots__ = ['schema', 'references', 'schema_type']
+
+    def __init__(self, schema, schema_type='AVRO', references=[]):
+        super(Schema, self).__init__()
+
+        self.schema = schema
+        self.schema_type = schema_type
+        self.references = references
+
+    def __eq__(self, other):
+        return str(self) == str(other)
+
+    def __str__(self):
+        return json.dumps(self.to_json(), sort_keys=True)
+
+    # TODO: consider metaclass and/or decorators
+    # JSON duck type handling
+    def to_json(self):
```

Comment:
as noted before, you're better off with just dumb data classes in scenarios like this (often, in general).

a couple of points:
1. the duck encoding thing you've got going results in a bit more indirection (=> a bit harder to follow)
2. it couples the schema class to the schema registry client class. this is not a big issue here of course, since they're basically coupled anyway, this is a general rule of thumb thing. throw OOP out the window. think functions and data, it's much better.